File: tournaments/smallestUnusualNumber/smallestUnusualNumber.cs
int smallestUnusualNumber(string a) {
    if (1 == a.Length) {
        return 10 - (a[0] - '0');
    }
    
    if (-1 != a.IndexOf('0') || isUnusual(a)) {
        return 0;        
    }
    
    return 10 - (a[a.Length - 1] - '0');
}

bool isUnusual(string s) {
    var index = 0;
    var n = s.Length;
    var sumOfOnes = 0;
    var sumOfRest = 0;
    var mulOfRest = 1;
    
    while ((mulOfRest - sumOfRest - sumOfOnes) < n && index < n) {
        var item = s[index++];
        if (1 == (item - '0')) {
            ++sumOfOnes;
        } else {
            sumOfRest += (item - '0');
            mulOfRest *= (item - '0');
        }
    }
    
    return (1 >= s.Length - sumOfOnes || (sumOfOnes + sumOfRest) > mulOfRest);
}

File: tournaments/smallestUnusualNumber/smallestUnusualNumber.js
function smallestUnusualNumber(a) {
    const isUnusual = function(s) {
        const n = s.length;
        let index = 0;
        let sumOfOnes = 0;
        let sumOfRest = 0;
        let mulOfRest = 1;

        while ((mulOfRest - sumOfRest - sumOfOnes) < n && index < n) {
            const item = s[index++];
            if (1 === (item.charCodeAt() - zeroCode)) ++sumOfOnes;
            else {
                sumOfRest += (item.charCodeAt() - zeroCode);
                mulOfRest *= (item.charCodeAt() - zeroCode);
            }
        }

        return (1 >= n - sumOfOnes || (sumOfOnes + sumOfRest) > mulOfRest);
    };

    const zeroCode = '0'.charCodeAt();
    const len = a.length;

    if (1 === len) return 10 - (a.charCodeAt(0) - zeroCode);
    if (~a.indexOf('0') || isUnusual(a)) return 0;
    return 10 - (a.charCodeAt(len - 1) - zeroCode);
}
